components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,319 @@
+  uint256_t value;
+  if (!GetUint256FromData(input, offset, value))
+    return false;
+
+  // Since we use value as an array index, we need to cast the type to
+  // something that can be used as an index, viz. size_t. To prevent runtime
+  // errors, we make sure the value is within safe limits of size_t.
+  if (value > std::numeric_limits<size_t>::max())
+    return false;
+
+  arg = static_cast<size_t>(value);
+  return true;
+}
+
+// GetBoolFromData extracts a 32-byte wide boolean value from the
+// calldata at the specified offset.
+//
+// The parsed bool value is serialized as "true" or "false" strings.
+bool GetBoolFromData(const std::string& input,
+                     size_t offset,
+                     std::string& arg) {  // NOLINT(runtime/references)
+  uint256_t value;
+  if (!GetUint256FromData(input, offset, value))
+    return false;
+
+  if (value == 0)
+    arg = "false";
+  else if (value == 1)
+    arg = "true";
+  else
+    return false;
+
+  return true;
+}
+
+// GetBytesHexFromData extracts a bytes value from the calldata at the
+// specified offset using head-tail encoding mechanism. bytes are packed
+// tightly in chunks of 32 bytes, with the first 32 bytes encoding the length,
+// followed by the actual content.
+//
+// The parsed bytearray is serialized as a hex string prefixed by "0x".
+bool GetBytesHexFromData(const std::string& input,
+                         size_t offset,
+                         std::string& arg) {  // NOLINT(runtime/references)
+  size_t pointer;
+  if (!GetSizeFromData(input, offset, pointer))
+    return false;
+
+  size_t bytes_len;
+  if (!GetSizeFromData(input, pointer * 2, bytes_len))
+    return false;
+
+  if (input.length() < pointer * 2 + 64 + bytes_len * 2)
+    return false;
+  arg = "0x" + input.substr(pointer * 2 + 64, bytes_len * 2);
+  return true;
+}
+
+// GetAddressArrayFromData parses a calldata sequence to extract a dynamic
+// array of addresses at the specified offset using head-tail encoding
+// mechanism. The encoding is similar to bytes, with the first 32 bytes
+// representing the number of elements in the array, followed by each element.
+//
+// The parsed data is joined together into a hex string prefixed by "0x".
+bool GetAddressArrayFromData(const std::string& input,
+                             size_t offset,
+                             std::string& arg) {  // NOLINT(runtime/references)
+  size_t pointer;
+  if (!GetSizeFromData(input, offset, pointer))
+    return false;
+
+  size_t array_len;
+  if (!GetSizeFromData(input, pointer * 2, array_len))
+    return false;
+
+  size_t array_offset = pointer * 2 + 64;
+  arg = "0x";
+  for (size_t i = 0; i < array_len; i++) {
+    std::string value;
+    if (!GetAddressFromData(input, array_offset, value))
+      return false;
+    base::StrAppend(&arg, {value.substr(2)});
+    array_offset += 64;
+  }
+
+  return true;
+}
+
+bool ABIDecodeInternal(const std::vector<std::string>& types,
```

Comment:
unit test coverage for ABIDecode is extremely insufficient, it only covers address.

Reply:
Added unit tests for all types here: 879e42f2a5f9136c1e6f65f803bd0954d1a88d6f. `eth_data_parser_unittest.cc` contains more unit tests with realistic transaction data.